cpp/cmake/3rdparty.cmake

```diff
@@ -0,0 +1,38 @@
+find_package(Git QUIET)
+
+if(NOT ${CODING_CONV_PREFIX}_THIRD_PARTY_DIR)
+  set(${CODING_CONV_PREFIX}_THIRD_PARTY_DIR 3rdparty)
+  set(${CODING_CONV_PREFIX}_THIRD_PARTY_DIR
+      3rdparty
+      PARENT_SCOPE)
+endif()
+
+# initialize submodule with given path
+function(bbp_init_git_submodule path)
+  if(NOT ${GIT_FOUND})
+    message(
+      FATAL_ERROR "git not found and ${path} submodule not cloned (use git clone --recursive)")
+  endif()
+  message(
+    STATUS "Fetching git submodule ${path}: running git submodule update --init --recursive ${path}"
+  )
+  execute_process(
+    COMMAND
+      ${GIT_EXECUTABLE} submodule update --init --recursive -- ${path}
+    WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
+    RESULT_VARIABLE git_submodule_status)
+  if(NOT git_submodule_status EQUAL 0)
+    message(FATAL_ERROR "Could not clone git submodule ${path}")
+  endif()
+endfunction()
+
+# check for external project and initialize submodule if it is missing
+function(bbp_add_git_submodule_subdirectory name)
+  set(submodule_path "${${CODING_CONV_PREFIX}_THIRD_PARTY_DIR}/${name}")
+  if(NOT EXISTS ${CMAKE_SOURCE_DIR}/${submodule_path}/CMakeLists.txt)
```

Comment:
CMAKE_SOURCE_DIR and not PROJECT_SOURCE_DIR?

Reply:
PROJECT_SOURCE_DIR is the path to the hpc-coding-conventions submodule.

Reply:
no, this is the path to the last `project()` and `CMAKE_SOURCE_DIR` is false if the project trying to use submodule is not the main project to be compiled.

Reply:
My bad, I thought the hpc-coding-conventions had its own `project`. Thank you for noticing this one, `CMAKE_{SOURCE,BINARY}_DIR` was everywhere (see dce9a67)